PEP draft for +T and -T syntax (#2)
jorenham/peps: pep-9999.rst

PEP: 9999
Title: Inline variance operators for generic type variables: ``+T`` and ``-T``
Author: Joren Hammudoglu <jhammudoglu at gmail.com>
Sponsor: Jelle Zijlstra <jelle.zijlstra at gmail.com>
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 24-Jul-2021
Post-History: 24-Jul-2021

Abstract
========

This PEP proposes a consise syntax for specifying the variance of
the type parameters of `user-defined generic types
<https://www.python.org/dev/peps/pep-0484/#user-defined-generic-types>`_.

> This allow declaring a co- or contravariant type parameter with ``+T`` or 

Reviewer: ```suggestion
This allows declaring a co- or contravariant type parameter with ``+T`` or 
```

``-T``, where ``T`` is a type variable.

Terminology
===========

This PEP uses some terms to refer to specific concepts related to Python 
typing, or type systems in general. For the sake of clarity, this PEP 

> considers the following terms and definitions:

Reviewer: This section looks scarier than it needs to be to readers: all of these terms are standard in typing, not specific to this PEP. But this section makes it look like you need PEP-specific definitions for all these terms.

It doesn't look like the Python docs have a glossary just yet, but that could be a useful addition.

Author: Do you think I should remove it alltogether, or restructure (a part of) it and rename the section to e.g. "introduction", or "background"?

Reviewer: I'd suggest renaming it to the bottom and removing the introductory paragraph.

type variable
    An instance of ``typing.TypeVar`` or ``typing.ParamSpec``, as defined
    in `PEP 484 <https://www.python.org/dev/peps/pep-0484/>`_ and 
    PEP `PEP 612 <https://www.python.org/dev/peps/pep-0612/>`_. Type 

>     variables can be used as either type annotations, or to declare 

Reviewer: ```suggestion
    variables can be used either as type annotations or to declare 
```

    user-defined generic types.

type parameter

>     Generic types can have one or multiple type parameters. For user-defined

Reviewer: I think you can drop "For user-defined generic types". TypeVars also used for builtin generics, the TypeVars are just only in typeshed.

    generic types, these are declared using type variables. Unlike type variables,
    type parameters only occur in generic types.

generic type
    A generic type, or generic class, is a class with a base class
    ``typing.Generic``, ``typing.Protocol``, or another generic type, 

>     and has at least one type parameter.

Reviewer: ```suggestion
    that has at least one type parameter.
```

    
variance
    A type parameter of a generic type can be either covariant, contravariant,
    or invariant (the default). This defines the subtyping relation of the 
    generic parameters and the generic type. See `PEP 484
    <https://www.python.org/dev/peps/pep-0484/#covariance-and-contravariance>`_.
    

Rationale
=========

> In PEP 484 it states:

Reviewer: ```suggestion
PEP 484 states:
```

    Covariance or contravariance is not a property of a type variable,
    but a property of a generic class defined using this variable. 

More specifically, variance is a property of the *type parameters* of a 
generic class. 

> Currently, variant type parameters are declared with type varianbles

Reviewer: ```suggestion
Currently, variant type parameters are declared with type variables
```

that are constructed with ``covariant=True`` or ``contravariant=True``, 
e.g.:
::

    from typing import TypeVar, Generic, Protocol
    
    T_co = TypeVar("T_co", covariant=True)
    T_contra = TypeVar("T_contra", contravariant=True)
    
    class SimpleFunction(Protocol[T_contra, T_co]):
        def __call__(self, x: T_contra) -> T_co: ...

But type variables can also be used outside of the scope of generic 
types, where the concept of variance does not apply, e.g.:
::

    def first(things: Iterable[T]) -> T:
        return next(iter(things))

This example is valid, but using ``T_co`` or ``T_contra`` instead of 
``T`` here does not make sense.

In ``SimpleFunction``, ``T_co`` or ``T_contra`` are used to annotate

> its method as well. This is required because the variance is associated 

Reviewer: ```suggestion
In ``SimpleFunction``, ``T_co`` and ``T_contra`` are used to annotate
a method. This is required because the variance is associated 
```

> with the individual type variables. But the method signature is unrelated 

Reviewer: I'm not sure this is true: the method signature does have something to do with the variance of the typevars, because the reason they need to be contravariant and covariant is because of the places where they appear in the signature.

Author: I think you're right @JelleZijlstra . I was trying to point out that associating variance with a type variable should only happen once, and that currently that isn't possible, which is pretty confusing. But I'm not sure whether it's technically incorrect to mark a typevar as variant within a method, or whether it's just redundant. What do you think? Should I leave this example out? And do you think it'd help illustrating how the current variance api is confusing by adding similar examples of how to use it incorrectly?

it makes more sense to specify the variance of type parameters directly
where the generic type is defined, removing the need to specify it on 
the type variables themselves. 

Proposal
========

The variance of the type parameters of a generic class can be specified
with the ``+`` and ``-`` prefix operators on their respective type
variables within the type parameter list of ``typing.Generic``, 
``typing.Protocol``, or the generic base class. Instead of
::

    T_co = typing.TypeVar('T_co', covariant=True)
    T_contra = typing.TypeVar('T_contra', contravariant=True)

    class SupportsInvertOld(typing.Protocol[T_co]):
        def __invert__(self) -> T_co: ...

    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
        def __le__(self, other: T_contra): ...

The new syntax uses the ``+`` and ``-`` prefix operators to specify
variant type parameters in the same place where the generic type is
declared:
::

    T = TypeVar("T")
    
    # covariant
    class SupportsInvert(typing.Protocol[+T]):
        def __invert__(self) -> T: ...

    # contravariant
    class SupportsPartialOrder(typing.Protocol[-T]):
        def __le__(self, other: T): ...

> This syntax is inspired by Scala programming language [1]_.

Reviewer: ```suggestion
This syntax is inspired by the Scala programming language [1]_.
```
You can also mention that this is already being used in the repr() of TypeVars.

Specification
=============

Valid use locations
-------------------

The new type variable variance syntax can only be used within the type
parameter list of ``typing.Generic``, ``typing.Protocol``, or a generic 
base class. 

The ``+`` and ``-`` prefix operators can be used on ``typing.TypeVar``
and ``typing.ParamSpec`` instances, iff 

- it is not already co- or contravariant

> - it has no type constraints 

Reviewer: The standard term is "value restriction" (https://mypy.readthedocs.io/en/stable/generics.html#type-variables-with-value-restriction)

Author: In PEP 484 that isn't mentioned, and refer to it as "constraints" instead. And in cpython, they are stored in `TypeVar`'s `__constraints__` attribute as a tuple. So wouldn't it make more sense to talk about "contraints" for `TypeVar(..., A, B)` and "bound(s)" for `TypeVar(..., bound=A)`?

Reviewer: Interesting, I guess mypy's doocs are the odd one out there.

> - it has no type bound

Reviewer: Variance is actually meaningful on TypeVars with a bound: https://mypy-play.net/?mypy=latest&python=3.10&gist=f72bac261a2148ca8ae6d6670dfb3459.

And similarly for value restrictions, see this example: https://mypy-play.net/?mypy=latest&python=3.10&gist=5a4f16dd13bd4b18a7c2323ba464f627. Mypy allows this but I think that's wrong; pyright gives an error on the last line. I opened https://github.com/python/mypy/issues/11379.

Reviewer 2: Variance with value restriction is counter-intuitive, and I suspect it might actually be best to forbid that.

PEP-484 uses the following example

```python
def concat(x: AnyStr, y: AnyStr) -> AnyStr:
    return x + y

class MyStr(str): ...

x = concat(MyStr('apple'), MyStr('pie'))
```

and says that the type bound to `AnyStr` would be `str` and not `MyStr`. This sounds exotic and tbh as a user I would be very confused by the fact that the type variable was bound to a type different from the type of the argument.

Author: @JelleZijlstra I see how it's useful for covariant types to have bounds or value restrictions. But a contravariant type with bounds is strange one. In PEP484 it's explicitly called an "upper bound", but that would make bounded contravariant types useless, since they can only "go up". So without lower bounds being a thing, I think that we either

- only allow covariant typevars to have bounds/value restrictions, 
- forbid bounds and restrictions on variant typevars allthogether for now,
- change the definitions so that contravariant typevars "flip" the bounds/restrictions,
- or leave it (explicitly) unspecified.

I think the last option would the cleanest, perhaps this issue will be more appropriate in e.g. your `T <= int` PEP @JelleZijlstra .

Reviewer: I think this PEP should stay purely syntactic and not change restrictions on when covariant + bounds/constraints is allowed. That will keep it more focused. If we want changes to what kind of semantics are allowed, that can be discussed separately.

PEP: 9999
Title: Inline variance operators for generic type variables: ``+T`` and ``-T``
Author: Joren Hammudoglu <jhammudoglu at gmail.com>
Sponsor: Jelle Zijlstra <jelle.zijlstra at gmail.com>
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 24-Jul-2021
Post-History: 24-Jul-2021

Abstract
========

This PEP proposes a consise syntax for specifying the variance of
the type parameters of `user-defined generic types
<https://www.python.org/dev/peps/pep-0484/#user-defined-generic-types>`_.
This allow declaring a co- or contravariant type parameter with ``+T`` or 
``-T``, where ``T`` is a type variable.

Terminology
===========

This PEP uses some terms to refer to specific concepts related to Python 
typing, or type systems in general. For the sake of clarity, this PEP 
considers the following terms and definitions:

type variable

>     An instance of ``typing.TypeVar`` or ``typing.ParamSpec``, as defined

Reviewer 2: I've just realized that `typing.ParamSpec` also allows to specify variance at construction time. Do you think we could avoid adding new syntax for `typing.ParamSpec`, given that the semantics of variance annotations is undefined?

Author: I don't think it matters that the variadic behaviour of `ParamSpec` is undefined at this moment, just like it is/will be for `TypeVarTuple` from PEP 646. Both inherit from `typing._TypeVarLike`  (although I'm guessing here for `TypeVarTuple`), where the `__covariant__` and `__contravariant__` attributes are set. So I think that `typing._TypeVarLike` is the best place to overload `__pos__` and `__neg__`, so that they return an e.g. `_TypeVarAlias(_TypeVarLike)` instance with an `__origin__: _TypeVarLike` attribute.

Reviewer 2: Apologies for bringing this up again, but I'm not sure implementation convenience is relevant here.

I think we should avoid adding syntax for features with unspecified behavior. Given that the variance of `typing.ParamSpec` variables is unspecified, I would say we should exclude them from this PEP. 

@JelleZijlstra wdyt?

Reviewer: `ParamSpec("P", covariant=True)` works at runtime, so I think `+P` should work at runtime too. Type checkers are free to reject it or give it an implementation-defined meaning.

    in `PEP 484 <https://www.python.org/dev/peps/pep-0484/>`_ and 
    PEP `PEP 612 <https://www.python.org/dev/peps/pep-0612/>`_. Type 
    variables can be used as either type annotations, or to declare 
    user-defined generic types.

type parameter
    Generic types can have one or multiple type parameters. For user-defined
    generic types, these are declared using type variables. Unlike type variables,
    type parameters only occur in generic types.

generic type
    A generic type, or generic class, is a class with a base class
    ``typing.Generic``, ``typing.Protocol``, or another generic type, 
    and has at least one type parameter.
    
variance
    A type parameter of a generic type can be either covariant, contravariant,
    or invariant (the default). This defines the subtyping relation of the 
    generic parameters and the generic type. See `PEP 484
    <https://www.python.org/dev/peps/pep-0484/#covariance-and-contravariance>`_.
    
variadic
    A variadic type parameter is either co- or contravariant. A variadic
    type (or variadic class) is a generic type with at least one variadic 
    type parameter
    

Rationale
=========

> In PEP 484 it states:

Reviewer 2: It's amusing that alongside that PEP-484 goes ahead and actually makes variance a property of a `typing.TypeVar`.

Author: I suppose it's one of the examples from python typing where the mindset "don't think about it too much, the type checker devs will figure out what to do with this" has lead to more confusion, defeating the purpose of improving readability and maintainability

    Covariance or contravariance is not a property of a type variable,
    but a property of a generic class defined using this variable. 

More specifically, variance is a property of the *type parameters* of a 
generic class. 

Currently, variadic type parameters are declared with type varianbles
that are constructed with ``covariant=True`` or ``contravariant=True``, 
e.g.:
::

    from typing import TypeVar, Generic, Protocol
    
    T_co = TypeVar("T_co", covariant=True)
    T_contra = TypeVar("T_contra", contravariant=True)
    
    class SimpleFunction(Protocol[T_contra, T_co]):
        def __call__(self, x: T_contra) -> T_co: ...

But type variables can also be used outside of the scope of generic 
types, where the concept of variance does not apply, e.g.:
::

    def first(things: Iterable[T]) -> T:
        return next(iter(things))

This example is valid, but using ``T_co`` or ``T_contra`` instead of 

> ``T`` here does not make sense, and type checkers do not allow is.

Reviewer 2: My experience is that type checkers ignore variance in most contexts where it does not apply, instead of reporting an error.

For example, all of the following definitions type check with mypy 0.902

```
def consume_one(thing: T_co) -> None: ...

def consume(things: Iterable[T_contra]) -> None: ...

Callback = Callable[[T_co], T_contra]
```

Author: I believe that (at least) mypy does not allow this

Reviewer 2: mypy does not allow a contravariant type variable in the return type annotation, but it does allow all of the above :)

In ``SimpleFunction``, ``T_co`` or ``T_contra`` are used to annotate
its method as well. This is required because the variance is associated 
with the individual type variables. But the method signature is unrelated 
to the variance of the type parameters of its generic type, 
``SimpleFunction`` in this example. 

Because variance is restricted to the type parameters of generic types, 
it makes more sense to specify the variance of type parameters directly
where the generic type is defined, removing the need to specify it on 
the type variables themselves. 

> 

Reviewer 2: Repeating my point from the email. 

The problem with existing variance API is that, although PEP-484 says otherwise, variance does look like a property of a type variable. This is confusing and requires users to mentally ignore variance annotations outside of generic class context. They could still write something like

```
def f(x: T_co) -> ...: ...
```

and that would look like a function with a covariant parameter, but in fact it isn't.

Author: It's strange that this isn't explicitly specified in PEP 484, but I believe that mypy does not allow this. Although you could make the case that `def f(x: T_contra) -> T_co: ...` is valid, because it is (approximately) equivalent to
```python
class F(Generic[T_contra, T_co]):
    def __call__(self, x: T_contra) -> T_co: ....

f = F()
```
But let's not overcomplicate things for now, and restrict the concept of variadic type parameters to generic types.

Reviewer 2: PEP-484 says that variance should be ignored outside of generic classes. You reference that note earlier in your PEP.

I might be wrong, but I think `F` is different from `f` in that the type arguments of `F` are fully known before `__call__` is called, so their variance is irrelevant for `__call__`, whereas the type arguments of `f` are inferred from the call.

Author: I think you're right, my example is indeed different in that aspect

Proposal
========

The variance of the type parameters of a generic class can be specified
with the ``+`` and ``-`` prefix operators on their respective type
variables within the type parameter list of ``typing.Generic``, 
``typing.Protocol``, or the generic base class. Instead of
::

    T_co = typing.TypeVar('T_co', covariant=True)
    T_contra = typing.TypeVar('T_contra', contravariant=True)

    class SupportsInvertOld(typing.Protocol[T_co]):
        def __invert__(self) -> T_co: ...

    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
        def __le__(self, other: T_contra): ...

The new syntax uses the ``+`` and ``-`` prefix operators to specify
variadic type parameters in the same place where the generic type is
declared:
::

    T = TypeVar("T")
    
    # covariant
    class SupportsInvert(typing.Protocol[+T]):
        def __invert__(self) -> T: ...

    # contravariant
    class SupportsPartialOrder(typing.Protocol[-T]):
        def __le__(self, other: T): ...

This syntax is inspired by Scala programming language [1]_.

> Specification

Reviewer 2: How should + and - work for a type variable with constraints? e.g.

```
T = TypeVar("T", str, bytes)
```

Author: I suppose that in this case it could make sense to use`T_A` for covariance, and `T_B` for contravariance.

```python
class A: ...
class B(A): ...

T_A = TypeVar("T_A", str, A)
T_B = TypeVar("T_B", str, B)
```

Reviewer 2: I'm sorry, I'm not sure what you're trying to say with this example.

My understanding of constraints is that `T` can only be substituted for `str` or `bytes` and not subtypes of either. This means that variance probably shouldn't apply to such type variables?

Reviewer: I don't know what variance would mean on a TypeVar with value restriction, so I'd be inclined to disallow it. But both mypy and pyright allow setting covariant=True on such a TypeVar, so maybe it means something I can't think of.

Regardless, it should be specified in the PEP. (And I also don't understand what the `T_A` example has to do with this topic.)

Author: The example does not make much sense in hindsight. I was trying to illustrate how a typevars are covariant in their upper bounds. Apparently this is used in Java on wildcard types ([wikipedia](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)#Use-site_variance_annotations_.28wildcards.29)), and in Scala, it is used when the type bound is another type parameter ([Scala docs](https://docs.scala-lang.org/tour/lower-type-bounds.html)).

In python, this would look something like `T = TypeVar('T', bound=+B)`, with `B` another typevar. But since this is not allowed at the moment, I agree that we shouldn't allow any constraints or bounds on variant typevars.

=============

Valid use locations
-------------------

The new type variable variance syntax can only be used within the type
parameter list of ``typing.Generic``, ``typing.Protocol``, or a generic 
base class. 

The ``+`` and ``-`` prefix operators can be used on ``typing.TypeVar``
and ``typing.ParamSpec`` instances, that are now already variadic, i.e.
its ``__covariant__`` and ``__contravariant__`` attributes must be 
``False``.

> When the same type variable is used on multiple generic base classes,

Reviewer 2: I guess the same constraint applies for the case of a single generic base class. E.g. the following should be rejected by the type checker

```
class A(Generic[T]): ...  # Invariant in T.
class B(A[+T]): ... 
```

because a covariant type parameter `T` is used in an invariant position in `A[...]`.

Author: I'm not sure that this is incorrect. At least, mypy allows it, and behaves as expected here:

```python
from typing import TypeVar, Generic

T = TypeVar('T')
T_co = TypeVar('T_co', covariant=True)
T_contra = TypeVar('T_contra', contravariant=True)

class Box(Generic[T]):
    value: T

    def __init__(self, value: T):
        self.value = value

class BoxCo(Box[T_co]): ...
class BoxContra(BoxCo[T_contra]): ...

class A: ...
class B(A): ...

def accepts_a(x: Box[A]) -> None: ...
def accepts_a_co(x: BoxCo[A]) -> None: ...
def accepts_a_contra(x: BoxContra[A]) -> None: ...
def accepts_b(x: Box[B]) -> None: ...
def accepts_b_co(x: BoxCo[B]) -> None: ...
def accepts_b_contra(x: BoxContra[B]) -> None: ...

a_box: Box[A] = Box(A())
a_box_co: BoxCo[A] = BoxCo(A())
a_box_contra: BoxContra[A] = BoxContra(A())

b_box: Box[B] = Box(B())
b_box_co: BoxCo[B] = BoxCo(B())
b_box_contra: BoxContra[B] = BoxContra(B())

accepts_a(a_box)
accepts_a_co(a_box_co)
accepts_a_contra(a_box_contra)

accepts_a(b_box)  # incompatible type "Box[B]"; expected "Box[A]"
accepts_a_co(b_box_co)
accepts_a_contra(b_box_contra)  # incompatible type "BoxContra[B]"; expected "BoxContra[A]"

accepts_b(a_box)  # incompatible type "Box[A]"; expected "Box[B]"
accepts_b_co(a_box_co)  # incompatible type "BoxCo[A]"; expected "BoxCo[B]"
accepts_b_contra(a_box_contra) 

accepts_b(b_box)
accepts_b_co(b_box_co)
accepts_b_contra(b_box_contra)
```

My guess is that defining `T` as covariant does not change the variance of `T` in A, i.e. even though the same type variable `T` is used to define the type parameters to `A` and `B`, the type parameters are independent from one another (not considering type constraints and bounds). 

But I can see that this could lead to issues, and that this should be discussed. Although I don't think that this PEP is the best place to impose additional restrictions on generic types, but should be left to type checkers, since type variance is not a concept that is specific to Python.

Reviewer 2: You're right, it is not as straightforward as I thought initially. 

I put together an example [demonstrating](https://mypy-play.net/?mypy=latest&python=3.10&gist=80be2a9951f8eff41f2a9ea3c2e1beee) unsoundness in my earlier comment -- the value of type `A[bool]` is also typed as `A[int]` in the same program, even though `A` is invariant in `T`. 

I would much prefer to settle this in a PEP instead of leaving it open to interpretation by tooling authors. Happy to start a thread on typing-sig@ if you feel this is best addressed in a separate PEP.

they must share the same variance, e.g.
::
    
    from typing import TypeVar, Callable, Container, Iterable, Protocol
    
    class LinkedList(Iterable[+T], Container[+T]): ...

>     class EventListener(Callable[[-T], None], Protocol[-T]): ...

Reviewer 2: I don't think you could use `Callable` in class bases. Consider changing to another type with contravariant parameters.

Author: Why not? It's an alias of `collections.abc.Callable`: https://docs.python.org/3/library/collections.abc.html#collections.abc.Callable

Reviewer 2: I might be behind on this, but PEP-585 doesn't say that `collections.abc.Callable` can be indexed in the same way as the `Callable` special-form. 

The typing docs, thought, do [say](https://docs.python.org/3.10/library/typing.html#typing.Callable) that

> A plain `Callable` is equivalent to `Callable[..., Any]`, and in turn to `collections.abc.Callable`.

which makes me think that you cannot use `Callable` in bases, but then `typing.Callable` is deprecated since 3.9 :man_shrugging:... 

As an extra data-point mypy 0.910 rejects your example module the new +/- syntax.

Author: You're right @superbobry , not only is `Callable` not a protocol, it's not even a valid base class apparently 🤔 ,  I assumed that it was a protocol with a `__call__`  method that has covariant arg and kwarg types, and a contravariant return type, but I suppose that it was wishful thinking.

    
are valid examples.

> Differences with current syntax

Reviewer 2: Given that all attributes here are implementation-defined, why talk about them at all?

Author: Is it? Are there other python implementations that do it differently?

Reviewer 2: Sorry, I'm not sure what you mean here.

My point was that this section describes the runtime representation of a `TypeVar` which is irrelevant for most tools consuming type information. Moreover, all referenced attributes have dunder names and are not part of the [public API](https://docs.python.org/3/library/typing.html#typing.TypeVar), which is why I thought it might be best to leave them out.

-------------------------------

> The new typevar operators return a transparent wrapper around the 

Reviewer 2: Does it make sense for `+T` and `-T` return the same type variable, so that variance is purely syntactic and is not a property of a type variable?

Author: I believe that all typing annotations should be accessible at runtime as well, and all are at the moment. These could be used by e.g. runtime type checkers (although I'm not sure if any of them currently support type variables).

Reviewer 2: Fair point. Returning `self` would make it impossible to get variance at runtime.

I would argue that type annotations shouldn't be accessible at runtime, but that ship has sailed :)

original type variable, which can be accessed with the ``__origin__``
attribute on the returned wrapper. e.g.::

    (+T).__origin__ is T
    (+T).__covariant__ is True
    (+T).__contravariant__ is False
    (+T).__name__ == T.__name__
    (+T).__constraints__ == T.__constraints__
    (+T).__bound__ is T.__bound__

> Thus, type variables defined with ``covariant=True`` and 

Reviewer 2: Why is it important that `+T` is not the same as `TypeVar("T", covariant=True)` at runtime?

Author: Because type variables must have unique names and should not be created at runtime.

Reviewer 2: I'm not sure I understand this. `+T` is conceptually a reference to the type variable named `"T"` and as such it doesn't introduce any new names. Could you elaborate?

Reviewer: Making it not equivalent to `TypeVar("T", covariant=True)` makes for a more straightforward implementation that is still also understandable for dynamic type checkers.

``contravariant=True``, are not equivalent to ``+T`` and ``-T``.

> ``+T`` and ``-T`` are not valid type annotations, and should only be

Reviewer 2: I suspect type checkers could enforce that on the syntax level without inspecting any of the dunder methods.

Author: I guess I'll see when I'll start on the reference implementation for mypy

Reviewer 2: Some type checkers are not written in Python and therefore cannot inspect the runtime state of type vars, e.g. pyre, pyright and PyCharm. These implementation details will be irrelevant for them. mypy of course is another story, but IIUC it also doesn't evaluate type annotations.

Reviewer: This section does feel a bit out of place in "Differences with current syntax".

used within the generic type parameter list of generic base classes, e.g.:: 

    class Spam(typing.Generic[+KT]): ...
    class Eggs(typing.Protocol[-KT, +VT]): ...
    class HamSet(typing.Sequence[+T]): ...

are valid uses.

All variance rules that apply to user-defined generic types should apply
in the same way with the new syntax, as they do with the current syntax,
and vice-versa.

> 

Reviewer 2: Should we outline a deprecation plan for variance parameters in `typing.TypeVar`?

Author: I don't think so. PEP 604 hasn't done so either.

Reviewer 2: My views are likely biased, since I work on tooling consuming type annotations, but I would much prefer for both PEP-604 and [PEP-613](https://www.python.org/dev/peps/pep-0613/) to outline a deprecation plan for the feature they improve on. Otherwise, tooling authors are left in a situation where they have to support multiple variations of a feature forever.

Happy to bring this up on typing-sig@ independently from this PEP.

Reviewer 2: @jorenham wdyt about explicitly saying that the variance parameters will not be deprecated and should remain a valid and supported way of declaring variance? 

I would personally prefer otherwise, but at least this leaves no ambiguity and clearly states that both the old and the new way should be supported by type checkers.

Rejected Ideas
==============

> For more detauls about discussions, see links below:

Reviewer: ```suggestion
For more details about discussions, see links below:
```
